File: frontend/src/services/product.service.ts
import apiClient from "@/lib/api-client";
import { CategoryDTO, ProductDTO } from "@/types";

export interface SearchParams {
  q?: string;
  category?: string;
  minPrice?: number;
  maxPrice?: number;
  promotionOnly?: boolean;
  sortBy?: string;
  sellerId?: number;
  page?: number;
  size?: number;
}

export interface PaginatedResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

export const productService = {
  async getProducts(params: SearchParams): Promise<PaginatedResponse<ProductDTO>> {
    const response = await apiClient.get<PaginatedResponse<ProductDTO>>("/products", { params });
    return response.data;
  },

  async getProductById(id: number): Promise<ProductDTO> {
    const response = await apiClient.get<ProductDTO>(`/products/${id}`);
    return response.data;
  },

  async getCategories(): Promise<CategoryDTO[]> {
    const response = await apiClient.get<CategoryDTO[]>("/categories");
    return response.data;
  },

  async getPromotionProducts(page = 0, size = 10): Promise<PaginatedResponse<ProductDTO>> {
    const response = await apiClient.get<PaginatedResponse<ProductDTO>>("/products/promotions", {
      params: { page, size }
    });
    return response.data;
  },

  async saveProduct(data: Partial<ProductDTO>): Promise<ProductDTO> {
    const response = await apiClient.post<ProductDTO>("/products", data);
    return response.data;
  },

  async updateProduct(id: number, data: Partial<ProductDTO>): Promise<ProductDTO> {
    const response = await apiClient.put<ProductDTO>(`/products/${id}`, data);
    return response.data;
  },

  async deleteProduct(id: number): Promise<void> {
    await apiClient.delete(`/products/${id}`);
  }
};

File: backend/src/main/java/com/shopflow/service/ProductService.java
package com.shopflow.service;

import com.shopflow.dto.ProductDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductService {
    Page<ProductDTO> getAllProducts(Pageable pageable);
    Page<ProductDTO> searchProducts(String q, String category, Double minPrice, Double maxPrice, Boolean promotionOnly, String sortBy, Long sellerId, Pageable pageable);
    Page<ProductDTO> getPromotionProducts(Pageable pageable);
    Page<ProductDTO> getTopSellingProducts(Pageable pageable);
    ProductDTO getProductById(Long id);
    ProductDTO saveProduct(ProductDTO productDTO, Long sellerId);
    ProductDTO updateProduct(Long id, ProductDTO productDTO, Long sellerId);
    void deleteProduct(Long id); // Sera un "soft delete"
}
